File: backend/app/api/v1/inventory_transactions.py
from fastapi import Depends, Query, status
from fastapi.routing import APIRouter

from app.auth.permissions import Permission, require_permission
from app.models.enums import SourceType, TransactionType
from app.schemas.auth import UserContext
from app.schemas.inventory_transactions import (
    DefectReportCreate,
    InventoryTransaction,
    InventoryTransactionList,
)
from app.services.inventory_transactions import InventoryTransactionService
from app.svc_dependencies import get_inv_transaction_svc

router = APIRouter()


@router.get("", response_model=InventoryTransactionList)
async def list_inventory_transactions(
    page: int = Query(1, ge=1),
    size: int = Query(10, ge=1, le=100),
    transaction_type: TransactionType | None = None,
    source_type: SourceType | None = None,
    _: UserContext = require_permission(Permission.INVENTORY_READ),
    service: InventoryTransactionService = Depends(get_inv_transaction_svc),
):
    txns, total = await service.list(
        page=page, size=size, transaction_type=transaction_type, source_type=source_type
    )

    return InventoryTransactionList(
        items=txns,
        total=total,
        size=size,
        page=page,
    )


@router.post(
    "/defects",
    response_model=InventoryTransaction,
    status_code=status.HTTP_201_CREATED,
)
async def report_defect(
    data: DefectReportCreate,
    current_user: UserContext = require_permission(Permission.INVENTORY_WRITE),
    service: InventoryTransactionService = Depends(get_inv_transaction_svc),
):
    return await service.report_defect(data=data, user=current_user)

File: frontend/src/api/inventory-transactions.ts
import client from "./client";
import type { InventoryTransaction, InventoryTransactionCreate, InventoryTransactionList } from "../types";
import { TransactionType, SourceType } from "../types";

export async function createInventoryTransaction(
  payload: InventoryTransactionCreate
): Promise<InventoryTransaction> {
  const { data } = await client.post<InventoryTransaction>("/v1/inventory-transactions", payload);
  return data;
}

export async function listInventoryTransactions(
  page: number,
  size: number,
  transaction_type?: TransactionType,
  source_type?: SourceType
): Promise<InventoryTransactionList> {
  const params: Record<string, unknown> = { page, size };
  if (transaction_type) params.transaction_type = transaction_type;
  if (source_type) params.source_type = source_type;
  const { data } = await client.get<InventoryTransactionList>("/v1/inventory-transactions", { params });
  return data;
}
